update: interopedia links (#4074)
JetBrains/kotlin-web-site: docs/topics/native/native-objc-interop.md

>
{type="warning"}

> This document covers some details of Kotlin/Native interoperability with Swift/Objective-C: how you can use Kotlin

Author: The problem is that it's not an exhaustive guide, we just highlight some aspects of the interop here

Can be called from Swift like this:

```swift
MyLibraryUtilsKt.foo()
```

See a collection of examples on accessing top-level Kotlin declarations in the Kotlin-Swift interopedia:

* [Top-level functions](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/overview/Top-level%20functions.md)
* [Top-level read-only properties](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/functionsandproperties/Top-level%20val%20properties.md)
* [Top-level mutable properties](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/functionsandproperties/Top-level%20mutable%20var%20properties.md)

### Method names translation

Generally, Swift argument labels and Objective-C selector pieces are mapped to Kotlin parameter names. These two concepts
have different semantics, so sometimes Swift/Objective-C methods can be imported with a clashing Kotlin signature.
In this case, the clashing methods can be called from Kotlin using named arguments, for example:

```swift
[player moveTo:LEFT byMeters:17]
[player moveTo:UP byInches:42]
```

In Kotlin, it's:

```kotlin
player.moveTo(LEFT, byMeters = 17)
player.moveTo(UP, byInches = 42)
```

> Here's how  some methods of `kotlin.Any` are mapped to Swift/Objective-C:

Reviewer: ```suggestion
Here's how functions of `kotlin.Any` are mapped to Swift/Objective-C:
```

As far as I can tell, these are all of the [functions](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-any/#functions) available in `kotlin.Any`. I thought to check because I kept reading the sentence as implying ”...and other functions may be mapped differently.”

| Kotlin       | Swift          | Objective-C   |
|--------------|----------------|---------------|
| `equals()`   | `isEquals(_:)` | `isEquals:`   |
| `hashCode()` | `hash`         | `hash`        |
| `toString()` | `description`  | `description` |

[See an example with data classes in the Kotlin-Swift interopedia](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/classesandinterfaces/Data%20classes.md).

You can specify a more idiomatic name in Swift or Objective-C, instead of renaming the Kotlin declaration with
the [`@ObjCName` annotation](#custom-declaration-names).

### Errors and exceptions

> Kotlin has no concept of checked exceptions, all Kotlin exceptions are unchecked. Swift has only checked errors.

Reviewer: This feels like a contradiction: if there is no concept of checked exceptions, they cannot be unchecked because unchecked is the opposite of checked. If that makes sense. In any case, I think this can be shortened to say, “All Kotlin exceptions are unchecked.”

Also, it follows that for people with a Kotlin-heavy background, it might be nice to include a short aside or a link about what checked exceptions are. Maybe a link to our [concepts](https://kotlinlang.org/docs/exceptions.html#checked-exceptions), but that page, too, mentions only that checked exceptions aren't there.

So if Swift or Objective-C code calls a Kotlin method which throws an exception to be handled, the Kotlin method
should be marked with the `@Throws` annotation specifying a list of "expected" exception classes.

When compiling to the Objective-C/Swift framework, non-`suspend` functions that have or inherit the `@Throws` annotation
are represented as `NSError*`-producing methods in Objective-C and as `throws` methods in Swift.
Representations for `suspend` functions always have`NSError*`/`Error` parameter in completion handler.

When Kotlin function called from Swift/Objective-C code throws an exception which is an instance of one of
the `@Throws`-specified classes or their subclasses, it is propagated as `NSError`.
Other Kotlin exceptions reaching Swift/Objective-C are considered unhandled and cause program termination.

> `suspend` functions without `@Throws` propagate only `CancellationException` as `NSError`.

Reviewer: ```suggestion
`suspend` functions without `@Throws` propagate only `CancellationException` (as `NSError`).
```

Without punctuation, this might read, “and other exceptions are propagated as something else.”

in Swift/Objective-C terminology.

Starting from Swift 5.5, Kotlin's `suspend` functions are also available for calling from Swift as
`async` functions without using the completion handlers. Currently, this functionality is highly experimental and

> has certain limitations. See [this YouTrack issue](https://youtrack.jetbrains.com/issue/KT-47610)for details.

Reviewer: ```suggestion
has certain limitations. See [this YouTrack issue](https://youtrack.jetbrains.com/issue/KT-47610) for details.
```

* Learn more about the [`async`/`await` mechanism in the Swift documentation](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html).
* See an example and recommendations on third-party libraries that implement the same functionality in the [Kotlin-Swift interopedia](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/coroutines/Suspend%20functions.md).

### Extensions and category members

Members of Objective-C categories and Swift extensions are generally imported to Kotlin as extensions. That's why

> these declarations can't be overridden in Kotlin. And the extension initializers aren't available as Kotlin constructors.

Reviewer: ```suggestion
these declarations can't be overridden in Kotlin, and the extension initializers aren't available as Kotlin constructors.
```

> 
{type="note"}

See more examples in the Kotlin-Swift interopedia:

* [How to access Kotlin objects using `shared`](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/classesandinterfaces/Objects.md)
* [How to access members of Kotlin companion objects from Swift](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/classesandinterfaces/Companion%20objects.md).

### NSNumber

> Kotlin primitive type boxes are mapped to special Swift/Objective-C classes. For example, `kotlin.Int` box is represented

Reviewer: ```suggestion
Kotlin primitive type boxes are mapped to special Swift/Objective-C classes. For example, the `kotlin.Int` box is represented
```

as `KotlinInt` class instance in Swift (or `${prefix}Int` instance in Objective-C, where `prefix` is the framework names prefix).
These classes are derived from `NSNumber`, so the instances are proper `NSNumber`s supporting all corresponding operations.

`NSNumber` type is not automatically translated to Kotlin primitive types when used as a Swift/Objective-C parameter type
or return value. The reason is that `NSNumber` type doesn't provide enough information about a wrapped primitive value
type, for example, `NSNumber` is statically not known to be `Byte`, `Boolean`, or `Double`. So Kotlin primitive values

> should be cast to and from `NSNumber` [manually below](#casting-between-mapped-types).

Reviewer: ```suggestion
should be [cast to and from `NSNumber` manually](#casting-between-mapped-types).
```

Or just remove the word "below," I guess.

### Collections

Kotlin collections are converted to Swift/Objective-C collections as described in the [table above](#mappings).
Swift/Objective-C collections are mapped to Kotlin in the same way, except for `NSMutableSet` and `NSMutableDictionary`.

> `NSMutableSet` isn't converted to a Kotlin `MutableSet`. To pass an object for Kotlin `MutableSet`, you can create this

Reviewer: This is a dense paragraph, mostly one huge sentence, with "create by creating" and two "or" constructions. It would be a smaller problem if the interopedia example had an example for `MutableSet`, but it doesn't as far as I can see.

Author: I've split and rephrased it a bit, hope it's better now

`KotlinMutableSet` class in Swift (or `${prefix}MutableSet` in Objective-C, where `prefix` is the framework names prefix).
The same holds for `MutableMap`.

[See an example in the Kotlin-Swift interopedia](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/overview/Collections.md).

### Function types

Kotlin function-typed objects (for example, lambdas) are converted to functions in Swift and blocks in Objective-C.
[See an example of a Kotlin function with a lambda in the Kotlin-Swift interopedia](https://github.com/kotlin-hands-on/kotlin-swift-interopedia/blob/main/docs/functionsandproperties/Functions%20returning%20function%20type.md).

However, there is a difference in how types of parameters and return values are mapped when translating a function and a
function type. In the latter case, primitive types are mapped to their boxed representation. Kotlin `Unit` return value
is represented as a corresponding `Unit` singleton in Swift/Objective-C. The value of this singleton can be retrieved the
same way as for any other Kotlin `object`. See singletons in the [table above](#mappings).

> So, the following Kotlin function:

Reviewer: ```suggestion
Consider the following Kotlin function:
```

```kotlin
fun foo(block: (Int) -> Unit) { ... }
```

> Is represented in Swift as:

Reviewer: ```suggestion
It is represented in Swift as follows:
```

```swift
func foo(block: (KotlinInt) -> KotlinUnit)
```

> And you can call it like:

Reviewer: ```suggestion
And you can call it like this:
```
